odbc/src/api/odbc_uri.rs

```diff
@@ -19,23 +38,75 @@ impl<'a> ODBCUri<'a> {
-        // the algorithm will most likely need to be a state machine over each character in the
-        // odbc_uri string.
-        odbc_uri
-            .split(';')
-            .filter(|attr| !attr.is_empty())
-            .map(|attr| {
-                // now split each attribute pair on '='
-                let mut sp = attr.split('=').collect::<Vec<_>>();
-                if sp.len() != 2 {
-                    return Err(ODBCError::InvalidUriFormat(EQUAL_ERROR.to_string()));
-                }
-                // ODBC attribute keys are case insensitive, so we lowercase the keys
-                Ok((sp.remove(0).to_lowercase(), sp.remove(0)))
-            })
-            .collect::<Result<HashMap<_, _>>>()
-            .map(ODBCUri)
+        let mut input = odbc_uri;
+        let mut ret = ODBCUri(HashMap::new());
+        while let Some((keyword, value, rest)) = ODBCUri::get_next_attribute(input)? {
+            ret.0.insert(keyword, value);
+            if rest.is_none() {
+                return Ok(ret);
+            }
+            input = rest.unwrap();
+        }
+        Ok(ret)
+    }
+
+    fn get_next_attribute(odbc_uri: &'a str) -> Result<Option<(String, &'a str, Option<&'a str>)>> {
+        // clean up any extra semi-colons
+        let index = odbc_uri.find(|c| c != ';');
+        // these are just trailing semis on the URI
+        if index.is_none() {
+            return Ok(None);
+        }
+        let odbc_uri = odbc_uri.get(index.unwrap()..).unwrap();
+        // find the first '=' sign, '=' does not appear in any keywords, so this is safe.
+        let (keyword, rest) = odbc_uri.split_at(
+            odbc_uri
+                .find('=')
+                .ok_or_else(|| ODBCError::InvalidUriFormat(EQUAL_ERROR.to_string()))?,
+        );
+        // remove the leading '=' sign.
+        let rest = rest.get(1..).unwrap();
+        if !KEYWORDS.is_match(keyword) {
+            return Err(ODBCError::InvalidUriFormat(format!(
+                "'{}' is not a valid URI keyword",
+                keyword
+            )));
+        }
+        let (value, rest) = if rest.starts_with('{') {
+            let rest = rest.get(1..).ok_or_else(|| {
+                ODBCError::InvalidUriFormat(MISSING_CLOSING_BRACE_ERROR.to_string())
+            })?;
+            ODBCUri::handle_braced_value(rest)?
+        } else {
+            ODBCUri::handle_unbraced_value(rest)?
+        };
+        Ok(Some((keyword.to_lowercase(), value, rest)))
+    }
+
+    fn handle_braced_value(input: &'a str) -> Result<(&'a str, Option<&'a str>)> {
+        let (value, rest) =
+            input.split_at(input.find('}').ok_or_else(|| {
+                ODBCError::InvalidUriFormat(MISSING_CLOSING_BRACE_ERROR.to_string())
+            })?);
+        match rest.len() {
+            0 => unreachable!(),
+            1 => return Ok((value, None)),
+            2 if rest.chars().nth(1).unwrap() == ';' => return Ok((value, None)),
```

Comment:
These two took me a reread to understand. Maybe we can add a comment like we have in the code above.

```suggestion
            // Just the "}" remaining
            1 => return Ok((value, None)),
            // Just "};" remaining
            2 if rest.chars().nth(1).unwrap() == ';' => return Ok((value, None)),
```
something like that ^